modules/light-clients/08-wasm/types/proposal_handle.go

```diff
@@ -10,34 +12,37 @@ import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
 
 	clienttypes "github.com/cosmos/ibc-go/v8/modules/core/02-client/types"
+	types "github.com/cosmos/ibc-go/v8/modules/core/02-client/types"
 	"github.com/cosmos/ibc-go/v8/modules/core/exported"
 )
 
-// CheckSubstituteAndUpdateState will try to update the client with the state of the
-// substitute.
-func (cs ClientState) CheckSubstituteAndUpdateState(
-	ctx sdk.Context,
-	_ codec.BinaryCodec,
-	subjectClientStore, substituteClientStore storetypes.KVStore,
-	substituteClient exported.ClientState,
-) error {
+// CheckSubstituteAndUpdateState will verify that a substitute client state is valid and update the subject client state.
+// Note that this method is used only for recovery and will not allow changes to the code hash.
+func (cs ClientState) CheckSubstituteAndUpdateState(ctx sdk.Context, cdc codec.BinaryCodec, subjectClientStore, substituteClientStore storetypes.KVStore, substituteClient exported.ClientState) error {
 	var (
 		subjectPrefix    = []byte("subject/")
 		substitutePrefix = []byte("substitute/")
 	)
 
-	_, ok := substituteClient.(*ClientState)
+	substituteClientState, ok := substituteClient.(*ClientState)
 	if !ok {
 		return errorsmod.Wrapf(
 			clienttypes.ErrInvalidClient,
-			fmt.Sprintf("invalid substitute client state. expected type %T, got %T", &ClientState{}, substituteClient),
+			fmt.Sprintf("invalid substitute client state. Expected type %T, got %T", &ClientState{}, substituteClient),
 		)
 	}
 
+	// check that code hashes of subject client state and substitute client state match
+	// changing the code hash is only allowed through the migrate contract RPC endpoint
+	if !bytes.Equal(cs.CodeHash, substituteClientState.CodeHash) {
+		return errorsmod.Wrapf(types.ErrInvalidClient, "expected code hashes to be equal: expected %s, got %s", hex.EncodeToString(cs.CodeHash), hex.EncodeToString(substituteClientState.CodeHash))
+	}
+
 	store := newUpdateProposalWrappedStore(subjectClientStore, substituteClientStore, subjectPrefix, substitutePrefix)
 
+	// new payload message
 	payload := SudoMsg{
-		CheckSubstituteAndUpdateState: &CheckSubstituteAndUpdateStateMsg{},
+		MigrateClientStore: &MigrateClientStoreMsg{},
```

Comment:
see [this pr ](https://github.com/misko9/centauri/pull/5/files)for the contract changes pertaining to this new msg type